### PR TITLE
[CDN-1233] Exception handling for get_certs_by_domain calls

### PR DESCRIPTION
**Pull request by coolyuvee**
The method get_certs_by_domain either returns an SSLCertificate object
if found else raises an exception otherwise. Whoever calling this
method should handle the raised exception.

Also, Updated the test cases to take care of these exception handling.

### CODE REVIEW

File: poppy/storage/cassandra/certificates.py [hunk 1]
```diff
@@ -248,27 +248,41 @@ def get_certs_by_domain(self, domain_name, project_id=None,
...
+        stmt = query.SimpleStatement(CQL_SEARCH_CERT_BY_DOMAIN,
             consistency_level=self._driver.consistency_level)
         self.session.row_factory = dict_factory
         result = self.session.execute(stmt, args)
 
         try:
-            return ssl_certificate.SSLCertificate.init_from_dict(result[0])
+            cert_obj = result[0]
+            for k, v in cert_obj.items():
+                if k == "cert_details":
```
Comment (heiths) [inline poppy/storage/cassandra/certificates.py]: This should be in a separate commit.

File: tests/unit/manager/default/test_background_job.py [hunk 2]
```diff
@@ -677,7 +677,7 @@ def test_delete_http_policy_cert_not_deployed(self):
 
     def test_delete_http_policy_cert_doesnt_exist(self):
         self.mock_storage.certificates_controller.\
-            get_certs_by_domain.return_value = []
+            get_certs_by_domain.side_effect = ValueError
```
Comment (heiths) [inline tests/unit/manager/default/test_background_job.py]: For this test (and other tests like it), it would be better to use assert raises decorator: `@raises(ValueError)`.

File: tests/unit/storage/cassandra/test_certificates.py [hunk 3]
```diff
@@ -65,17 +65,54 @@ def setUp(self):
...
+        self.mock_session.execute.return_value = cert_details_json[0]
+        ssl_cert = self.cc.get_certs_by_domain( 'www.mydomain.com',
+                                               flavor_id = 'flavor1',
+                                               cert_type = 'san')
         self.assertTrue(isinstance(ssl_cert, ssl_certificate.SSLCertificate))
-        self.mock_session.execute.assert_called_with(ANY, args)
+        self.assertEqual(ssl_cert.flavor_id, 'flavor1')
+        self.assertEqual(ssl_cert.project_id, '12345')
+        self.assertEqual(ssl_cert.cert_type, 'san')
+        self.mock_session.execute.assert_called()
```
Comment (heiths) [inline tests/unit/storage/cassandra/test_certificates.py]: Why not `assert_called_with(SQL_QUERY_HERE)`? Since we know exactly what the SQL statement should look like with a given domain.

File: poppy/manager/default/ssl_certificate.py [hunk 4]
```diff
@@ -430,14 +435,8 @@ def get_certs_by_status(self, status):
...
     def update_certificate_status(self, domain_name, certificate_updates):
-        certificate_old = self.storage.get_certs_by_domain(domain_name)
-        if not certificate_old:
-            raise ValueError(
-                "certificate information not found for {0} ".format(
-                    domain_name
-                )
-            )
 
+        certificate_old = self.storage.get_certs_by_domain(domain_name)
```
Comment (heiths) [inline poppy/manager/default/ssl_certificate.py]: Does this not need a "try/except"?

File: poppy/storage/cassandra/certificates.py [hunk 5]
```diff
@@ -248,27 +248,41 @@ def get_certs_by_domain(self, domain_name, project_id=None,
             'domain_name': domain_name.lower(),
         }
 
-        query_string  = CQL_SEARCH_CERT_BY_DOMAIN
```
Comment (heiths) [inline poppy/storage/cassandra/certificates.py]: I wonder how much of these additional params are actually necessary...  Can a domain be associated with multiple flavors, cert_types or projects?

If a domain exists in the `certificate_info` table, there are validators in place to prevent adding another of the same name.  In fact, domain_name is the primary key for the table.  

If that's the case -- this could be greatly simplified.

File: poppy/manager/default/ssl_certificate.py (hunk 4 above)
Comment (coolyuvee) [reply poppy/manager/default/ssl_certificate.py]: No. Earlier the code was explicitly throwing a ValueError when there is no certificate found for the given domain name. Now, the method get_certs_by_domain() itself does the same thing implicitly.

File: poppy/storage/cassandra/certificates.py (hunk 5 above)
Comment (coolyuvee) [reply poppy/storage/cassandra/certificates.py]: Yes. There can only be a single record for any given domain name. Same domain name with other flavors/cert_types/projects will not be stored. Thats the reason only domain_name is part of the sql query.

File: tests/unit/storage/cassandra/test_certificates.py (hunk 3 above)
Comment (coolyuvee) [reply tests/unit/storage/cassandra/test_certificates.py]: Yes. Agree.

File: poppy/storage/cassandra/certificates.py (hunk 1 above)
Comment (coolyuvee) [reply poppy/storage/cassandra/certificates.py]: If it is in a separate commit .. The new tests will fail? We can discuss.

File: tests/unit/manager/default/test_background_job.py (hunk 2 above)
Comment (coolyuvee) [reply tests/unit/manager/default/test_background_job.py]: Could do that if the test is written only to call one method and is expected to receive an exception. However, several methods are called (or mocked for return values/exceptions). Will try and change wherever this is possible(assert exceptions)
